spring-integration-jms/src/main/java/org/springframework/integration/jms/JmsOutboundGateway.java

```diff
@@ -726,34 +742,44 @@ protected Object handleRequestMessage(final Message<?> message) {
 					return null;
 				}
 			}
-			Object result = jmsReply;
-			if (this.extractReplyPayload) {
-				result = this.messageConverter.fromMessage(jmsReply);
-				if (logger.isDebugEnabled()) {
-					logger.debug("converted JMS Message [" + jmsReply + "] to integration Message payload [" + result + "]");
-				}
-			}
-			Map<String, Object> jmsReplyHeaders = this.headerMapper.toHeaders(jmsReply);
 
-			if (this.replyContainer != null && this.correlationKey != null) {
-				// do not propagate back the gateway's internal correlation id
-				jmsReplyHeaders.remove(this.correlationKey);
-			}
-			Message<?> replyMessage = null;
-			if (result instanceof Message){
-				replyMessage = this.getMessageBuilderFactory().fromMessage((Message<?>) result).copyHeaders(jmsReplyHeaders).build();
+			if (reply instanceof javax.jms.Message) {
+				return replyToReplyMessage((javax.jms.Message) reply);
 			}
 			else {
-				replyMessage = this.getMessageBuilderFactory().withPayload(result).copyHeaders(jmsReplyHeaders).build();
+				return reply;
 			}
-			return replyMessage;
 		}
 		catch (JMSException e) {
 			throw new MessageHandlingException(requestMessage, e);
 		}
 	}
 
-	private javax.jms.Message sendAndReceiveWithContainer(Message<?> requestMessage) throws JMSException {
+	private Message<?> replyToReplyMessage(javax.jms.Message jmsReply) throws JMSException {
+		Object result = jmsReply;
+		if (this.extractReplyPayload) {
+			result = this.messageConverter.fromMessage(jmsReply);
+			if (logger.isDebugEnabled()) {
+				logger.debug("converted JMS Message [" + jmsReply + "] to integration Message payload [" + result + "]");
+			}
+		}
+		Map<String, Object> jmsReplyHeaders = this.headerMapper.toHeaders(jmsReply);
+
+		if (this.replyContainer != null && this.correlationKey != null) {
+			// do not propagate back the gateway's internal correlation id
+			jmsReplyHeaders.remove(this.correlationKey);
+		}
+		Message<?> replyMessage = null;
+		if (result instanceof Message){
+			replyMessage = this.getMessageBuilderFactory().fromMessage((Message<?>) result).copyHeaders(jmsReplyHeaders).build();
+		}
+		else {
+			replyMessage = this.getMessageBuilderFactory().withPayload(result).copyHeaders(jmsReplyHeaders).build();
```

Comment:
We introduced the `MutableMessage` explicitly for this purpose (when a message is rebuilt several times within the framework as it is passed back up the call stack) but we never took the next step and actually used it for that purpose.